### PR TITLE
Scoreboard

### PR DESCRIPTION
**Pull request by SkyDiverStar**
# Description of changes

Simple scoreboard for now. Upgrades and changes can be made later. 
grafik
Screenshot 2023-07-19 165703

## How has this been tested?

Please describe the test strategy you followed.

- [ ] automated unit test
- [ ] automated integration test
- [ ] automated acceptance test
- [x] manual, exploratory test

In case of manual test, please document the test well including a set of user instructions and prerequisites. Each including an action, it's result, and where appropriate a screenshot.

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My code fulfilles all acceptance criteria
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have made corresponding changes to the documentation
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer

- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: app/courses/[courseId]/scoreboard/page.tsx [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  scoreboard.forEach((element, index: number) => {
+    index++;
+    rows.push(
+      createData(
+        `${index}`,
+        capitalizeFirstLetter(element.user.userName),
+        element.powerScore
+      )
+    );
+  });
```
Comment (v-morlock) [inline app/courses/[courseId]/scoreboard/page.tsx:92]: das würde ich eher so lösen:

```suggestion
  const rows: Data[] = scoreboard.map((element, index) => createData(
        `${index}`,
        capitalizeFirstLetter(element.user.userName),
        element.powerScore
      ));
```

File: app/courses/[courseId]/scoreboard/page.tsx [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        <TableContainer sx={{ maxHeight: "100%" }}>
+          <Table stickyHeader aria-label="sticky table">
+            <TableHead>
+              <TableRow>
+                {columns.map((column) => (
+                  <TableCell
+                    key={column.id}
+                    align={column.align}
+                    style={{ minWidth: column.minWidth }}
+                    sx={{ backgroundColor: "#90a4ae" }}
```
Comment (v-morlock) [inline app/courses/[courseId]/scoreboard/page.tsx:111]: ich würde überlegen ob wir dein custom-styling der tabelle hier und weiter unten rausnehmen, finde das ungestylte von material schicker um ehrlich zu sein:

image

Comment (v-morlock) [reply app/courses/[courseId]/scoreboard/page.tsx:111]: evtl sogar das "Paper" drum rum weglassen?

image

File: app/courses/[courseId]/scoreboard/page.tsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    );
+  });
+
+  return (
+    <div>
+      <Link href={`/courses/${courseId}`}>
+        <Button variant="contained" sx={{ margin: "10px" }}>
+          Back
+        </Button>
+      </Link>
```
Comment (v-morlock) [inline app/courses/[courseId]/scoreboard/page.tsx:100]: würde hier noch eine überschrift reinpacken und vlt den button zum zurückgehen nicht ganz so prominent? sowas wie hier von marcel?

image

File: app/courses/[courseId]/student.tsx [hunk 4]
```diff
@@ -70,11 +93,33 @@ export default function StudentCoursePage() {
...
   }
 
+  // Extract scoreboard
+  scoreboard.forEach((element, index: number) => {
+    if (index < 3) {
+      rows.push(createData(element.user.userName, element.powerScore));
+    }
+    index++;
+  });
+
```
Comment (v-morlock) [inline app/courses/[courseId]/student.tsx:122]: würde hier auch mit "map" arbeiten wie oben geschrieben

File: app/courses/[courseId]/student.tsx [hunk 5]
```diff
@@ -103,9 +148,42 @@ export default function StudentCoursePage() {
...
+                    </TableCell>
+                    <TableCell align="right">{row.power}</TableCell>
+                  </TableRow>
+                ))}
+              </TableBody>
+            </Table>
+          </TableContainer>
+          <Link href={{ pathname: `${id}/scoreboard` }}>
+            <Button variant="contained">Full Scoreboard</Button>
+          </Link>
```
Comment (v-morlock) [inline app/courses/[courseId]/student.tsx:185]: den würd ich auch nicht als primary button machen

File: app/courses/[courseId]/student.tsx [hunk 6]
```diff
@@ -103,9 +148,42 @@ export default function StudentCoursePage() {
...
+                  </TableCell>
+                </TableRow>
+              </TableHead>
+              <TableBody>
+                {rows.map((row) => (
+                  <TableRow
+                    key={row.name}
+                    sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
+                  >
+                    <TableCell component="th" scope="row">
```
Comment (v-morlock) [inline app/courses/[courseId]/student.tsx:174]: generell fände ichs vom layout cool wenn die linke box und die rechte box irgendwie ähnlich wären und auch von der ausrichtung oben aligned sind

image

File: app/courses/[courseId]/scoreboard/page.tsx (hunk 2 above)
Comment (SkyDiverStar) [reply app/courses/[courseId]/scoreboard/page.tsx:111]: Fun fact: Ich habe von der MUI Seite einfach den Code für "Sticky Header" tables kopiert. Ist also das MUI styling. 
Aber ja, kann ich machen

File: app/courses/[courseId]/student.tsx (hunk 6 above)
Comment (SkyDiverStar) [reply app/courses/[courseId]/student.tsx:174]: Die sollten jetzt beide oben gleich aligned sein. Aber mit dem Layout bin ich mir nicht sicher, was ich da machen soll. Finde es nicht schlecht wie es derzeit aussieht

Comment (SkyDiverStar) [reply app/courses/[courseId]/student.tsx:174]: Besonders, mein Plan war es die Top drei aus dem Kurs darzustellen + man selbst, falls man nicht in den Top 3 ist. Hab es aber erst mal hinten angestellt

Comment (v-morlock) [reply app/courses/[courseId]/student.tsx]: so hab ich gemeint

image

Comment (v-morlock) [reply app/courses/[courseId]/student.tsx]: passt denke ich kannst wegen mir mergen
